### PR TITLE
feat: add POST endpoint for tools

### PR DESCRIPTION
**Pull request by kushagra189**
Implementation of a `POST /tools` endpoint that stores data that should be sufficient to comply with `GET /tools/` and related endpoints. Needs more work for supporting all functionality.

Addresses #6

### CODE REVIEW

File: trs_filer/api/20200404.0aa5443.openapi.yaml [hunk 1]
```diff
@@ -661,7 +661,7 @@ components:
           description: The URL for this tool in this registry.
         id:
           type: string
-          example: 123456
+          example: "123456"
```
Comment (uniqueg) [inline trs_filer/api/20200404.0aa5443.openapi.yaml]: Bit stupid that they supplied an `int` as an example for a `str`, but did this actually cause any problems? If not, I'd rather prefer not to modify the original specs.

File: trs_filer/config.yaml [hunk 2]
```diff
@@ -52,3 +57,8 @@ exceptions:
     required_members: [['message'], ['code']]
     status_member: ['code']
     exceptions: trs_filer.errors.exceptions.exceptions
+
+endpoints:
+    objects:
```
Comment (uniqueg) [inline trs_filer/config.yaml]: This probably shouldn't be `objects` because that's the DRS name for its records. Here, it should probably rather be `tools`.

File: trs_filer/ga4gh/trs/server.py [hunk 3]
```diff
@@ -109,3 +113,18 @@ def toolClassesGet():
 def getServiceInfo():
     """ Show information about this service. """
     return {}
+
+
+@log_traffic
+def addTool() -> Dict:
+    tool_create_instance = CreateToolPostObject(request)
+    tool = tool_create_instance.create_object()
+    ret_tool = {
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/server.py]: I think it's cleaner to return from your helper class already the object as you want to return it. That is, put this reshaping/subsetting code in `.create_object()` or some other helper method.

And  no need to assign something to a variable if you only use that variable in one other place. For example here, there's no reason to add the overhead of assigning to a variable with `ret_tool = my_code()` and `return ret_tool` if you could just as well do `return my_code()`.

File: trs_filer/ga4gh/trs/server.py [hunk 4]
```diff
@@ -109,3 +113,18 @@ def toolClassesGet():
 def getServiceInfo():
     """ Show information about this service. """
     return {}
+
+
+@log_traffic
+def addTool() -> Dict:
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/server.py:122]: Please add a docstring.

File: trs_filer/ga4gh/trs/server.py [hunk 5]
```diff
@@ -1,8 +1,12 @@
 """"Controllers for TRS endpoints."""
 
 import logging
-
+from typing import Dict
+from flask import (request)
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/server.py]: No need for parentheses if you're only importing a single item

File: trs_filer/api/additional_post_endpoints.openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            schema:
+              x-body-name: tool
+              $ref: '#/components/schemas/ToolPost'
+      responses:
+        '200':
+          description: tool response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ToolPost'
```
Comment (uniqueg) [inline trs_filer/api/additional_post_endpoints.openapi.yaml]: I think your response should be just a single string with the ID of the created tool (see DRS-filer), not the object that you put in. Apart from all the extra stuff that is unnecessarily transmitted, without the identifier, the user wouldn't have a chance to (easily) find the tool they just registered.

File: trs_filer/api/additional_post_endpoints.openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        default:
```
Comment (uniqueg) [inline trs_filer/api/additional_post_endpoints.openapi.yaml]: I think the explicit error definition as is done in the other specs (WES, TES, DRS) is clearer. I think you can copy it straight from the DRS-filer additional specs, just with different descriptions. I think it should cover `400`, `401`, `403` and `500`. I know they don't have them in TRS, but since we can design the `POST` however we want, I think we can go with a more explicit definition.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
+"""Controller for adding new objects."""
+
+from random import choice
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Please mind the import order defined in [PEP8](https://www.python.org/dev/peps/pep-0008/#imports). For example, here it should be ordered:

```py
from random import choice
from typing import Dict

from flask import (current_app, request)
from pymongo.errors import DuplicateKeyError

from trs_filer.app import logger
```

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+from typing import Dict
+from trs_filer.app import logger
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Not so good. It is convention to create a logger for each module with:

```
import logging
# your other imports

logger = logging.getLogger(__name__)
```

Do _not_ import loggers from other modules.

Please check the [logging documentation](https://docs.python.org/3/library/logging.html) for more details why that is recommended.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+from flask import (current_app, request)
+from pymongo.errors import DuplicateKeyError
+
+
+class CreateToolPostObject:
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Don't forget to add docstrings for your class and methods.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    def __init__(self, request: request):
+        self.db_collection = (
+            current_app.config['FOCA'].db.dbs['trsStore'].
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: It's really a minor point, but I think adding the line break _before_ the dot is much cleaner, as it clearly indicates that we are calling a method in that second line (and not an independent call).

So:

```py
(
    my_long_line_object
    .method()
)
```

rather than

```py
(
    my_long_line_object.
    method()  # looks like a function call at first glance
)
```

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+from random import choice
+from typing import Dict
+from trs_filer.app import logger
+from flask import (current_app, request)
+from pymongo.errors import DuplicateKeyError
+
+
+class CreateToolPostObject:
+    def __init__(self, request: request):
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: I like the class-based approach here :) I think it makes some sense in the way you use it for data persistence

Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Please explicitly mention that you return `None` type. Always type the returns, as well as all params.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            collections['objects'].client
+        )
+        self.tool_data = request.json
+        self.id_charset = current_app.config['FOCA'].\
+            endpoints['objects']['id_charset']
+        self.id_length = current_app.config['FOCA'].\
+            endpoints['objects']['id_length']
+        self.host_name = current_app.config['FOCA'].server.host
+
+    def create_id(self, charset, length) -> str:
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Please add type info and full docstring. You can check DRS-filer

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        """Create tool class."""
+        self.tool_data['toolclass'] = {
+            "description": "Temporary tool class.",
+            "id": "123456",
+            "name": "ToolClass"
+        }
+
+    def create_object(self) -> Dict:
+        """
+        Add new tool post objects to TRS Registry.
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Too complicated! `Register tool.` or `Register tool with TRS.` will be clearer and more concise.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        if "checker_url" in self.tool_data:
+            self.tool_data['has_checker'] = True
+        else:
+            self.tool_data['has_checker'] = False
+
+        # set tool class (proper method needs to be added)
+        self.create_tool_class()
+
+        # set meta_version method
+
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Please implement a semantic version parser, but also allow user to override this behavior as discussed recently.

File: trs_filer/api/additional_post_endpoints.openapi.yaml [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+          type: string
+          description: The name of the tool.
+        description:
+          type: string
+          description: The description of the tool.
+        checker_url:
+          type: string
+          description: Optional url to the checker tool that will exit successfully if
+            this tool produced the expected result given test data.
+        versions:
```
Comment (uniqueg) [inline trs_filer/api/additional_post_endpoints.openapi.yaml]: Please re-add `meta_version` to allow the user to optionally set it him/herself

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        # set unique id and url and save object
+        while True:
+            try:
+                generated_object_id = self.\
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Please don't wrap your lines like that, it's hard to read and error-prone. Especially here, there is absolutely no need to do it, because you can simply put your method params on different lines. Also, you con merge this line with the next (because you use the `generated_object_id` only once, when assigning it to something else). Finally, for added readability/maintainability, it's always good to be explicit, so if a function/method has named arguments, it's nice to _always_ call them by name, too. Then, if somebody rearranges the args in `.create_id()`, the call will still work.

This is better:

```py
                self.tool_data['id'] = self.create_id(
                    charset=self.id_charset,
                    length=self.id_length,
                )
```

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+                    create_id(self.id_charset, self.id_length)
+                self.tool_data['id'] = generated_object_id
+                self.tool_data['url'] = \
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Use parantheses to wrap lines, not `\`. Check DRS-filer

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+                    f"{self.host_name}/tools/{self.tool_data['id']}"
+                self.db_collection.insert_one(self.tool_data)
+            except DuplicateKeyError:
+                continue
+            except Exception as e:
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Here you are catching the expression, just to raise it again. What's the point? :) Just remove this. Also, since you are really only doing something particular when a `DuplicateKeyError` occurs, there is no need to put anything other than the call that can actually raise this error in the `try:` block, i.e., only the call to `.insert_one()` should be in there.

Please have a look at DRS-filer.

**Review by uniqueg — CHANGES_REQUESTED** [review]
The important stuff looks good (except for the response object in the `POST` endpoint specs, see comment), but there's quite a few very minor things, most of which arising from copying DRS-filer stuff that I had already fixed there in the final versions. Please work through the comments, shouldn't take that long...

File: trs_filer/api/20200404.0aa5443.openapi.yaml (hunk 1 above)
Comment (kushagra189) [reply trs_filer/api/20200404.0aa5443.openapi.yaml]: Not causing problems, tried a few cases. Should be fine , I think.

Comment (uniqueg) [reply trs_filer/api/20200404.0aa5443.openapi.yaml]: If it _doesn't_ cause problems without the double quotes around `123456`, could you please revert to the original (i.e., without the quotes), as this is apparently what is written in the original specs?

File: trs_filer/ga4gh/trs/server.py [hunk 20]
```diff
@@ -109,3 +112,10 @@ def toolClassesGet():
 def getServiceInfo():
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/server.py]: Add typing info for return value (`Dict`)

Comment (uniqueg) [reply trs_filer/ga4gh/trs/server.py]: Add this for the other endpoints (or stubs) as well.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py (hunk 10 above)
Comment (uniqueg) [reply trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Hate to be picky about this, but there should be empty lines between sections in Google-style docstrings.

Here's an example:

```py
def my_func(arg_1: str = "something") -> str:
    """Summary comment.

    Additional comments on usage etc., if applicable.

    Args:
        arg_1: Description of argument.

    Returns:
        Modified string.

    Raises:
        TypeError: User supplied argument of wrong type.
    """
    if isinstance(arg_1, str):
        return 2 * arg_1
    else:
        raise TypeError("Supplied argument is not of type `str`")
```

File: trs_filer/config.yaml [hunk 21]
```diff
@@ -13,13 +13,18 @@ db:
...
+                              id: 1
+                          options:
+                            'unique': True
 
 api:
     specs:
         - path:
             - api/20200404.0aa5443.openapi.yaml
             - api/20200131.11d2c12.service-info.yaml
+            - api/additional_post_endpoints.openapi.yaml
```
Comment (uniqueg) [inline trs_filer/config.yaml]: maybe just call it `additions.openapi.yaml`, otherwise you would need to change it once you put in some endpoint that is not a `POST`

File: trs_filer/api/additional_post_endpoints.openapi.yaml [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        description:
+          type: string
+          description: The description of the tool.
+        checker_url:
+          type: string
+          description: Optional url to the checker tool that will exit successfully if
+            this tool produced the expected result given test data.
+        meta_version:
+          type: string
+          description: The version of this tool in the registry. Iterates when fields like
```
Comment (uniqueg) [inline trs_filer/api/additional_post_endpoints.openapi.yaml]: Would be better to describe the implementation's behavior here, i.e., something like

> Unless provided, it is attempted to parse the latest version from the `meta_version` properties of the available `versions`. Unset if the latest version cannot be determined.

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+    ) -> str:
+        """Sort and give the latest version from version list.
+        Args:
+            lst_versions: List of versions of the tool.
+        Returns:
+            Latest version string from the list
+            if semantic versioning else empty string.
+        """
+        try:
+            lst_versions.sort(key=lambda s: list(map(int, s.split('.'))))
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: This will break if versions contain pre-release (https://semver.org/#spec-item-9) or build (https://semver.org/#spec-item-10) information. Please add support for these versions in the following way:

- strip all build information, as by the SemVer specs it doesn't have any bearing on precedence
- if there are only pre-releases, compare the part _before_ the hyphen numerically, like you do in the current implementation
- if there are regular releases AND pre-releases, compare only the stable releases

File: trs_filer/ga4gh/trs/endpoints/registerObjects.py [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+            return lst_versions[-1]
+        except Exception as e:
```
Comment (uniqueg) [inline trs_filer/ga4gh/trs/endpoints/registerObjects.py]: Would be better to catch one or more specific exceptions. Also, I would log this with `logger.warning()` and add the input list in the log message. Always write log messages as informative as possible, as they will help us make sure that everything runs well (and help us debug if it _doesn't_ work as expected)

**Review by uniqueg — CHANGES_REQUESTED** [review]
Few more small changes, but getting there ;-)

File: setup.cfg [hunk 25]
```diff
@@ -0,0 +1,6 @@
+[coverage:run]
+source =
+    trs_filer
+omit =
+    trs_filer/app.py
+    trs_filer/errors/exceptions.py
```
Comment (uniqueg) [inline setup.cfg]: Why did you add this? I suppose it shouldn't be too difficult to test?

**Review by uniqueg — CHANGES_REQUESTED** [review]
Looks good. Just wondering why you omit `exceptions.py` from testing
